Support Implicit Grant Authorization Flow (Fix #499)

IdmFoundInHim opened the pull request:
The code heavily borrows from the existing user authentication class, and the tests are slightly-modified copies of the SpotifyOAuth tests. All tests passed on 3.7.7

stephanebruckert commented on CHANGELOG.md:
```diff
@@ -9,6 +9,15 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
+   to refresh the token without user input. (However, read the class
+   docstring for security advisory.)
+
+## [2.12.1] - 2020-05-28
```
2.12.1 has not been released yet and I think we will release your PR as part of 2.13.0

For now just put your changes into [Unreleased]

IdmFoundInHim commented:
The failing check is a link in `help(SpotifyImplicitGrant)`. Is there a good way to ignore it, or should I shorten the link?

stephanebruckert commented:
@IdmFoundInHim about the line too long, just add `  # noqa: E501` 
https://stackoverflow.com/a/47876831/1515819

stephanebruckert commented:
amazing contribution! can't wait to test it, hopefully soon

stephanebruckert commented on spotipy/oauth2.py:
```diff
@@ -505,10 +506,248 @@ def _add_custom_values_to_token_info(self, token_info):
+
+    This auth manager enables *user and non-user* endpoints with only
+    a client secret, redirect uri, and username. The user will need to
+    copy and paste a URI from the browser every hour.
```
what is the main reason that the ability to refresh the token without user input is sacrificed? Is there a blocker that makes it technically impossible currently or is it just not implemented, and can still be implemented in the future?

IdmFoundInHim replied:
The blocker is that the token comes back in a hash fragment (instead of a query string). I could not get the server to retrieve the fragment

stephanebruckert commented on spotipy/oauth2.py:
```diff
@@ -505,10 +506,248 @@ def _add_custom_values_to_token_info(self, token_info):
+
+        Parameters
+        ----------
+        * as_dict: **DO NOT USE**, included for tests compatability
```
I can't find what is using this in the tests, could we not just delete it? I would rather not keep stuff that's only needed in tests.

compatability -> compatibility

IdmFoundInHim replied:
It was for `prompt_for_user_token`. Once `prompt_for_user_token` was deprecated, I switched to the new template, so it should be removable.

stephanebruckert commented:
I was about to merge this but since we also merged https://github.com/plamere/spotipy/pull/509 there are now conflicts, any chance you take a look at it?

IdmFoundInHim commented:
It's going to take some more work to resolve the conflicts without breaking the new feature. It may change the behavior of `Spotify Oauth.parse_oauth_response_url`...what are the plans for releasing v2.13 (deadlines, pending features, requirements)?

stephanebruckert commented:
Don't worry about 2.13.0, I might release it soon but I will create 2.14.0 just for your PR, so please take your time!

IdmFoundInHim commented:
I fixed the conflicts and added state verification for the new auth flow. Along the way, I tweaked the behavior of some of the new functions, the biggest changes being:
* Removed state verification from `prompt_for_user_token` since it is deprecated, and
* Adding a dedicated subclass for state mismatches (`SpotifyStateError`) because it is a potential security warning

stephanebruckert commented:
Nice work, merged!

stephanebruckert commented:
and released in [2.13.0](https://github.com/plamere/spotipy/releases/tag/2.13.0)
